samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.samza.context.Context;
+import org.apache.samza.operators.functions.ClosableFunction;
+import org.apache.samza.operators.functions.InitableFunction;
+import org.apache.samza.serializers.Serde;
+
+
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient CouchbaseEnvironment env;
+  protected transient Cluster cluster;
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
+  protected Long timeout = 0L;
+  protected TimeUnit timeUnit = null;
+  protected Integer ttl = 0; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected final List<String> clusterNodes;
+  protected final String bucketName;
+
+  // Environment Settings
+  protected CouchbaseEnvironmentConfigs environmentConfigs;
+
+  public BaseCouchbaseTableFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
+    Preconditions.checkArgument(StringUtils.isNotEmpty(bucketName), "Bucket name is not allowed to be null or empty.");
+    Preconditions.checkArgument(CollectionUtils.isNotEmpty(clusterNodes),
+        "Cluster nodes is not allowed to be null or empty.");
+    Preconditions.checkArgument(valueClass != null, "Value class is not allowed to be null.");
+    this.bucketName = bucketName;
+    this.clusterNodes = ImmutableList.copyOf(clusterNodes);
+    this.environmentConfigs = new CouchbaseEnvironmentConfigs();
+  }
+
+  @Override
+  public void init(Context context) {
+    bucket = COUCHBASE_BUCKET_REGISTRY.getBucket(bucketName, clusterNodes, environmentConfigs);
+  }
+
+  @Override
+  public void close() {
+    COUCHBASE_BUCKET_REGISTRY.closeBucket(bucketName, clusterNodes);
+  }
+
+  public boolean isRetriable(Throwable exception) {
+    while (exception != null && !(exception instanceof TemporaryFailureException)
+        && !(exception instanceof TemporaryLockFailureException)) {
+      exception = exception.getCause();
+    }
+    return exception != null;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTimeout(long timeout, TimeUnit timeUnit) {
+    this.timeout = timeout;
+    this.timeUnit = timeUnit;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTtl(int ttl) {
+    this.ttl = ttl;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withSerde(Serde<V> valueSerde) {
+    this.valueSerde = valueSerde;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withUsernameAndPassword(String username, String password) {
+    environmentConfigs.username = username;
+    environmentConfigs.password = password;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withSslEnabled(boolean sslEnabled) {
```

Comment:
Should this be enabled by default?

Reply:
Although the official doc says sslEnabled is true by default. In Java SDK, it's false in default. The problem is, if we set sslEnabled to be true, then sslKeystoreFile or sslTruststoreFile must be provided. So I think keeping it false by default makes sense.

Reply:
In practice, SSL is likely always enabled. Are you sure user need to provide additional parameters, can you confirm this?

Reply:
Yes. Enabling SSL here means using certificate-based authentication, which also means role-based authentication is disabled. https://github.com/couchbase/couchbase-jvm-core/blob/master/src/main/java/com/couchbase/client/core/env/DefaultCoreEnvironment.java#L1217-L1223